File: Q/Q3_code.R
# Q3_code.R
# Workshop: Iteration with purrr in R
# Prof: Timothy Fraser

# Now that we know how to write functions and use loops, let's learn a more 
# elegant way to apply functions to multiple items: the purrr package!
# 
# The purrr package provides powerful tools for iteration (repeating operations).
# Instead of writing for loops, we can use map() functions to apply a function
# to each element in a list or vector. This is often cleaner and easier to read.


# 1. Getting Started #######################################

# We'll need the purrr package for iteration functions
# and dplyr for data manipulation
library(purrr)
library(dplyr)

# We'll also use the built-in mtcars dataset
# This dataset contains information about 32 different car models
# Let's take a quick look at it
head(mtcars)
# You'll see columns like mpg (miles per gallon), cyl (cylinders), 
# hp (horsepower), etc.


# 2. What is Iteration? #######################################

# Iteration means repeating an operation on multiple items.
# 
# For example, imagine you want to calculate the mean of several columns
# in a dataset. Instead of writing:
#   mean(mtcars$mpg)
#   mean(mtcars$hp)
#   mean(mtcars$wt)
#   ... (and so on for each column)
#
# You can use iteration to apply the mean() function to each column automatically!
#
# The purrr package provides map() functions that make this easy.


# 3. Understanding map() #######################################

# map() applies a function to each element in a list or vector.
# The basic structure is:
#   map(.x = your_data, .f = your_function)
#
# - .x is the data you want to iterate over (a vector, list, or data frame columns)
# - .f is the function you want to apply to each element
#
# map() returns a list with the results


# 4. Your First map() #######################################

# Example: Calculate the mean of each numeric column in mtcars

# First, let's select just the numeric columns
numeric_cols = mtcars %>%
  select(where(is.numeric))

# Now use map() to calculate the mean of each column
means = map(.x = numeric_cols, .f = mean)

# Let's see what we got
means
# Output: A list where each element is the mean of one column
# $mpg: [1] 20.09062
# $cyl: [1] 6.1875
# $disp: [1] 230.7219
# ... and so on

# You can access individual results like this:
means$mpg
# Output: [1] 20.09062

means$hp
# Output: [1] 146.6875


# 5. map() with Different Functions #######################################

# You can use map() with any function!

# Example: Calculate the standard deviation of each column
standard_deviations = map(.x = numeric_cols, .f = sd)
standard_deviations

# Example: Find the minimum value in each column
minimums = map(.x = numeric_cols, .f = min)
minimums

# Example: Find the maximum value in each column
maximums = map(.x = numeric_cols, .f = max)
maximums


# 6. map() with Custom Functions #######################################

# You can also use map() with your own custom functions!

# Let's create a function that calculates the range (max - min)
calculate_range = function(x){
  max_value = max(x)
  min_value = min(x)
  range_value = max_value - min_value
  return(range_value)
}

# Now apply it to each column
ranges = map(.x = numeric_cols, .f = calculate_range)
ranges

# You can also write the function inline using ~ and . notation
# The ~ creates a function, and . represents each element
ranges = map(.x = numeric_cols, .f = ~ max(.) - min(.))
ranges


# 7. map() with Multiple Arguments #######################################

# Sometimes you want to pass additional arguments to your function.
# You can do this using additional arguments after .f

# Example: Calculate quantiles (percentiles) for each column
# The quantile() function takes a 'probs' argument to specify which quantiles
quantiles = map(.x = numeric_cols, .f = quantile, probs = c(0.25, 0.5, 0.75))
quantiles
# This calculates the 25th, 50th (median), and 75th percentiles for each column


# 8. map_dfr() - Combining Results into a Data Frame #######################################

# Often, you want your results in a nice data frame instead of a list.
# That's where map_dfr() comes in! The 'dfr' stands for "data frame rows."
# It combines all the results into a single data frame.

# Example: Calculate summary statistics for each column and combine into a data frame

# First, let's create a function that returns multiple statistics
get_summary_stats = function(x){
  data.frame(
    mean = mean(x),
    median = median(x),
    sd = sd(x),
    min = min(x),
    max = max(x)
  )
}

# Now apply it to each column using map_dfr()
# We need to also pass the column name, so let's modify our approach
summary_stats = map_dfr(.x = numeric_cols, .f = get_summary_stats, .id = "column")
summary_stats
# Output: A data frame with one row per column, showing all the statistics
# The .id = "column" argument creates a column showing which original column each row came from


# 9. A Better Approach: Using names() with map_dfr() #######################################

# Let's get the column names and create a more informative summary

# Create a function that takes both the data and the column name
get_summary_with_name = function(column_name, data){
  x = data[[column_name]]  # Get the column data
  data.frame(
    column = column_name,
    mean = mean(x),
    median = median(x),
    sd = sd(x),
    min = min(x),
    max = max(x)
  )
}

# Apply to each column name
column_names = names(numeric_cols)
summary_stats = map_dfr(.x = column_names, .f = get_summary_with_name, data = mtcars)
summary_stats
# Now we have a nice data frame with the column name and all statistics!


# 10. Practical Example: Analyzing Multiple Groups #######################################

# Let's use map_dfr() to analyze cars by number of cylinders

# Split the data by number of cylinders
cars_by_cyl = split(mtcars, mtcars$cyl)
# This creates a list where each element is a data frame of cars with that many cylinders

# Create a function to summarize each group
summarize_group = function(group_data){
  data.frame(
    cylinders = unique(group_data$cyl),
    n_cars = nrow(group_data),
    avg_mpg = mean(group_data$mpg),
    avg_hp = mean(group_data$hp),
    avg_wt = mean(group_data$wt)
  )
}

# Apply to each group
group_summaries = map_dfr(.x = cars_by_cyl, .f = summarize_group)
group_summaries
# Output: A data frame showing statistics for 4-cylinder, 6-cylinder, and 8-cylinder cars


# 11. map() vs. Loops #######################################

# You might be wondering: why use map() instead of a for loop?
# 
# Both work, but map() has advantages:
# 1. More concise and readable
# 2. Less error-prone (no need to manage indices)
# 3. Works well with the tidyverse style of coding
# 4. Automatically handles different data types
#
# However, for loops are still useful when you need:
# - Complex control flow (multiple conditions, breaks, etc.)
# - Operations that depend on previous iterations
# - When you're more comfortable with loops!

# Here's the same operation using a for loop (for comparison):
summary_list = list()
for(i in 1:length(column_names)){
  col_name = column_names[i]
  summary_list[[i]] = get_summary_with_name(col_name, mtcars)
}
summary_stats_loop = bind_rows(summary_list)
# This does the same thing, but map_dfr() is cleaner!


# 12. Other map() Variants #######################################

# purrr provides several variants of map() for different output types:
#
# - map() - returns a list
# - map_dbl() - returns a numeric vector (double)
# - map_int() - returns an integer vector
# - map_chr() - returns a character vector
# - map_lgl() - returns a logical (TRUE/FALSE) vector
# - map_dfr() - returns a data frame (combines rows)
# - map_dfc() - returns a data frame (combines columns)

# Example: Get means as a numeric vector instead of a list
mean_vector = map_dbl(.x = numeric_cols, .f = mean)
mean_vector
# Output: A named numeric vector
# mpg      cyl     disp       hp     drat       wt     qsec       vs       am     gear     carb 
# 20.09062  6.1875 230.7219 146.6875  3.596563  3.21725 17.84875  0.4375  0.40625  3.6875  2.8125


# 13. Cleaning Up #######################################

# Clear your environment
rm(list = ls())


# Conclusion #######################################

# Great! You've learned how to:
# - Use map() to apply functions to each element in a list or vector
# - Use map_dfr() to combine results into a data frame
# - Pass additional arguments to functions in map()
# - Use custom functions with map()
# - Understand when to use map() vs. loops
#
# The purrr package makes iteration cleaner and more powerful than traditional loops.
# Practice using map() and map_dfr() to make your code more elegant and efficient!

File: Q/Q3_code.py
# Q3_code.py
# Workshop: Iteration with map() in Python
# Prof: Timothy Fraser

# Now that we know how to write functions and use loops, let's learn a more 
# elegant way to apply functions to multiple items: Python's built-in map() function!
# 
# Python's map() function provides powerful tools for iteration (repeating operations).
# Instead of writing for loops, we can use map() to apply a function
# to each element in a list or sequence. This is often cleaner and easier to read.


# 1. Getting Started #######################################

# We'll need pandas for data manipulation
import pandas as pd

# We'll create a sample dataset similar to mtcars
# This dataset contains information about different car models
# Let's create a simple version with key columns
mtcars = pd.DataFrame({
    'mpg': [21.0, 21.0, 22.8, 21.4, 18.7, 18.1, 14.3, 24.4, 22.8, 19.2, 17.8, 16.4, 17.3, 15.2, 10.4, 10.4, 14.7, 32.4, 30.4, 33.9, 21.5, 15.5, 15.2, 13.3, 19.2, 27.3, 26.0, 30.4, 15.8, 19.7, 15.0, 21.4],
    'cyl': [6, 6, 4, 6, 8, 6, 8, 4, 4, 6, 6, 8, 8, 8, 8, 8, 8, 4, 4, 4, 4, 8, 8, 8, 8, 4, 4, 4, 8, 6, 8, 4],
    'disp': [160.0, 160.0, 108.0, 258.0, 360.0, 225.0, 360.0, 146.7, 140.8, 167.6, 167.6, 275.8, 275.8, 275.8, 472.0, 460.0, 440.0, 78.7, 75.7, 71.1, 120.1, 318.0, 304.0, 350.0, 400.0, 79.0, 120.3, 95.1, 351.0, 145.0, 301.0, 121.0],
    'hp': [110, 110, 93, 110, 175, 105, 245, 62, 95, 123, 123, 180, 180, 180, 205, 215, 230, 66, 52, 65, 97, 150, 150, 245, 175, 66, 91, 113, 264, 175, 335, 109],
    'drat': [3.90, 3.90, 3.85, 3.08, 3.15, 2.76, 3.21, 3.69, 3.92, 3.92, 3.92, 3.07, 3.07, 3.07, 2.93, 3.00, 3.23, 4.08, 4.93, 4.22, 3.70, 2.76, 3.15, 3.73, 3.08, 4.08, 4.43, 3.77, 4.22, 3.62, 3.54, 4.11],
    'wt': [2.620, 2.875, 2.320, 3.215, 3.440, 3.460, 3.570, 3.190, 3.150, 3.440, 3.440, 4.070, 3.730, 3.780, 5.250, 5.424, 5.345, 2.200, 1.615, 1.835, 2.465, 3.520, 3.435, 3.840, 3.845, 1.935, 2.140, 1.513, 3.170, 2.770, 3.570, 2.780],
    'qsec': [16.46, 17.02, 18.61, 19.44, 17.02, 20.22, 15.84, 20.00, 22.90, 18.30, 18.90, 17.40, 17.60, 18.00, 17.98, 17.82, 17.42, 19.47, 18.52, 19.90, 20.01, 16.87, 17.30, 15.41, 17.05, 18.90, 16.70, 16.90, 14.50, 15.50, 14.60, 18.60],
    'vs': [0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1],
    'am': [1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1],
    'gear': [4, 4, 4, 3, 3, 3, 3, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 4, 4, 4, 3, 3, 3, 3, 3, 4, 5, 5, 5, 5, 5, 4],
    'carb': [4, 4, 1, 1, 2, 1, 4, 2, 2, 4, 4, 3, 3, 3, 4, 4, 4, 1, 2, 1, 1, 2, 2, 4, 2, 1, 2, 2, 4, 6, 8, 2]
})

# Let's take a quick look at it
print(mtcars.head())
# You'll see columns like mpg (miles per gallon), cyl (cylinders), 
# hp (horsepower), etc.


# 2. What is Iteration? #######################################

# Iteration means repeating an operation on multiple items.
# 
# For example, imagine you want to calculate the mean of several columns
# in a dataset. Instead of writing:
#   mtcars['mpg'].mean()
#   mtcars['hp'].mean()
#   mtcars['wt'].mean()
#   ... (and so on for each column)
#
# You can use iteration to apply the mean() function to each column automatically!
#
# Python's built-in map() function makes this easy.


# 3. Understanding map() #######################################

# map() applies a function to each element in a sequence (list, tuple, etc.).
# The basic structure is:
#   map(function, sequence)
#
# - function is the function you want to apply to each element
# - sequence is the data you want to iterate over (a list, tuple, etc.)
#
# map() returns a map object (an iterator) with the results
# You can convert it to a list to see the results


# 4. Your First map() #######################################

# Example: Calculate the mean of each numeric column in mtcars

# First, let's select just the numeric columns
numeric_cols = mtcars.select_dtypes(include=['number'])

# Now use map() to calculate the mean of each column
# We'll iterate over the column names and apply mean to each column
column_names = numeric_cols.columns
means = {col: numeric_cols[col].mean() for col in column_names}

# Let's see what we got
print(means)
# Output: A dictionary where each key is a column name and value is the mean
# {'mpg': 20.090625, 'cyl': 6.1875, 'disp': 230.721875, ...}

# You can access individual results like this:
means['mpg']
# Output: 20.090625

means['hp']
# Output: 146.6875

# Alternative: Using map() with a list of column names
# This is closer to the R purrr style
def get_mean(col_name):
    return numeric_cols[col_name].mean()

means_list = list(map(get_mean, column_names))
# This gives a list of means (but loses column names)

# Better: Using dictionary comprehension (Pythonic way)
means = {col: numeric_cols[col].mean() for col in column_names}


# 5. map() with Different Functions #######################################

# You can use map() with any function!

# Example: Calculate the standard deviation of each column
standard_deviations = {col: numeric_cols[col].std() for col in column_names}
print(standard_deviations)

# Example: Find the minimum value in each column
minimums = {col: numeric_cols[col].min() for col in column_names}
print(minimums)

# Example: Find the maximum value in each column
maximums = {col: numeric_cols[col].max() for col in column_names}
print(maximums)


# 6. map() with Custom Functions #######################################

# You can also use map() with your own custom functions!

# Let's create a function that calculates the range (max - min)
def calculate_range(x):
    max_value = max(x)
    min_value = min(x)
    range_value = max_value - min_value
    return range_value

# Now apply it to each column using map()
ranges = {col: calculate_range(numeric_cols[col]) for col in column_names}
print(ranges)

# You can also write the function inline using a lambda function
# Lambda functions are anonymous functions defined with lambda
ranges = {col: (lambda x: max(x) - min(x))(numeric_cols[col]) for col in column_names}
print(ranges)

# Or more simply:
ranges = {col: numeric_cols[col].max() - numeric_cols[col].min() for col in column_names}
print(ranges)


# 7. map() with Multiple Arguments #######################################

# Sometimes you want to pass additional arguments to your function.
# You can do this by creating a wrapper function or using lambda with additional args

# Example: Calculate quantiles (percentiles) for each column
# The quantile() method in pandas takes a 'q' argument to specify which quantiles
def get_quantiles(col_name, probs):
    return numeric_cols[col_name].quantile(q=probs)

# Apply to each column
quantiles = {col: get_quantiles(col, [0.25, 0.5, 0.75]) for col in column_names}
print(quantiles)
# This calculates the 25th, 50th (median), and 75th percentiles for each column


# 8. Combining Results into a DataFrame #######################################

# Often, you want your results in a nice data frame instead of a dictionary.
# We can create a DataFrame from a list of dictionaries.

# Example: Calculate summary statistics for each column and combine into a data frame

# First, let's create a function that returns multiple statistics
def get_summary_stats(col_name):
    col_data = numeric_cols[col_name]
    return {
        'column': col_name,
        'mean': col_data.mean(),
        'median': col_data.median(),
        'sd': col_data.std(),
        'min': col_data.min(),
        'max': col_data.max()
    }

# Now apply it to each column using map() and convert to DataFrame
summary_list = [get_summary_stats(col) for col in column_names]
summary_stats = pd.DataFrame(summary_list)
print(summary_stats)
# Output: A data frame with one row per column, showing all the statistics


# 9. A Better Approach: Using column names directly #######################################

# Let's create a more informative summary using the column names directly

# Create a function that takes the column name and returns summary stats
def get_summary_with_name(column_name):
    col_data = numeric_cols[column_name]
    return {
        'column': column_name,
        'mean': col_data.mean(),
        'median': col_data.median(),
        'sd': col_data.std(),
        'min': col_data.min(),
        'max': col_data.max()
    }

# Apply to each column name
summary_stats = pd.DataFrame([get_summary_with_name(col) for col in column_names])
print(summary_stats)
# Now we have a nice data frame with the column name and all statistics!


# 10. Practical Example: Analyzing Multiple Groups #######################################

# Let's analyze cars by number of cylinders

# Split the data by number of cylinders using groupby
cars_by_cyl = {cyl: group for cyl, group in mtcars.groupby('cyl')}
# This creates a dictionary where each key is a cylinder count and value is a DataFrame

# Create a function to summarize each group
def summarize_group(group_data):
    return {
        'cylinders': group_data['cyl'].iloc[0],  # Get unique value (all same in group)
        'n_cars': len(group_data),
        'avg_mpg': group_data['mpg'].mean(),
        'avg_hp': group_data['hp'].mean(),
        'avg_wt': group_data['wt'].mean()
    }

# Apply to each group
group_summaries = pd.DataFrame([summarize_group(group) for group in cars_by_cyl.values()])
print(group_summaries)
# Output: A data frame showing statistics for 4-cylinder, 6-cylinder, and 8-cylinder cars


# 11. map() vs. Loops #######################################

# You might be wondering: why use map() or list comprehensions instead of a for loop?
# 
# Both work, but map() and list comprehensions have advantages:
# 1. More concise and readable
# 2. Less error-prone (no need to manage indices)
# 3. More Pythonic (idiomatic Python style)
# 4. Automatically handles different data types
#
# However, for loops are still useful when you need:
# - Complex control flow (multiple conditions, breaks, etc.)
# - Operations that depend on previous iterations
# - When you're more comfortable with loops!

# Here's the same operation using a for loop (for comparison):
summary_list = []
for col_name in column_names:
    summary_list.append(get_summary_with_name(col_name))
summary_stats_loop = pd.DataFrame(summary_list)
# This does the same thing, but list comprehensions are cleaner!


# 12. Other Iteration Methods #######################################

# Python provides several ways to iterate and apply functions:
#
# - map() - returns a map object (iterator)
# - List comprehensions - returns a list directly (often preferred in Python)
# - Dictionary comprehensions - returns a dictionary
# - Generator expressions - memory-efficient iteration
#
# Example: Get means as a dictionary (we already did this above)
mean_dict = {col: numeric_cols[col].mean() for col in column_names}
# Wait, that's a set comprehension! For dict, we need:
mean_dict = {col: numeric_cols[col].mean() for col in column_names}
print(mean_dict)
# Output: A dictionary with column names as keys and means as values


# 13. Cleaning Up #######################################

# Clear your environment
# In Python, you can use globals().clear() to clear all global variables
# But be careful - this removes everything!

# globals().clear()  # Uncomment to clear all variables


# Conclusion #######################################

# Great! You've learned how to:
# - Use map() and list comprehensions to apply functions to each element
# - Combine results into a DataFrame
# - Pass additional arguments to functions in iterations
# - Use custom functions with map() and comprehensions
# - Understand when to use map() vs. loops
#
# Python's map() function and list comprehensions make iteration cleaner and more powerful than traditional loops.
# Practice using map() and comprehensions to make your code more elegant and efficient!

